x-pack/plugin/core/src/test/java/org/elasticsearch/xpack/core/ilm/CreateSnapshotStepTests.java

```diff
@@ -0,0 +1,169 @@
+                        is("snapshot name was not generated for policy [" + policyName + "] and index [" + indexName + "]"));
+                }
+            });
+        }
+
+        {
+            IndexMetaData.Builder indexMetadataBuilder =
+                IndexMetaData.builder(indexName).settings(settings(Version.CURRENT).put(LifecycleSettings.LIFECYCLE_NAME, policyName))
+                    .numberOfShards(randomIntBetween(1, 5)).numberOfReplicas(randomIntBetween(0, 5));
+            IndexMetaData indexMetaData = indexMetadataBuilder.build();
+
+            ClusterState clusterState =
+                ClusterState.builder(emptyClusterState()).metaData(MetaData.builder().put(indexMetaData, true).build()).build();
+
+            CreateSnapshotStep createSnapshotStep = createRandomInstance();
+            createSnapshotStep.performAction(indexMetaData, clusterState, null, new AsyncActionStep.Listener() {
+                @Override
+                public void onResponse(boolean complete) {
+                    fail("expecting a failure as the index doesn't have any snapshot name in its ILM execution state");
+                }
+
+                @Override
+                public void onFailure(Exception e) {
+                    assertThat(e, instanceOf(IllegalStateException.class));
+                    assertThat(e.getMessage(),
+                        is("snapshot repository is not present for policy [" + policyName + "] and index [" + indexName + "]"));
+                }
+            });
+        }
+    }
+
+    public void testPerformAction() {
+        String indexName = randomAlphaOfLength(10);
+        String policyName = "test-ilm-policy";
+        Map<String, String> ilmCustom = new HashMap<>();
+        String snapshotName = indexName + "-" + policyName;
+        ilmCustom.put("snapshot_name", snapshotName);
+        String repository = "repository";
+        ilmCustom.put("snapshot_repository", repository);
+
+        IndexMetaData.Builder indexMetadataBuilder =
+            IndexMetaData.builder(indexName).settings(settings(Version.CURRENT).put(LifecycleSettings.LIFECYCLE_NAME, policyName))
+                .putCustom(LifecycleExecutionState.ILM_CUSTOM_METADATA_KEY, ilmCustom)
+                .numberOfShards(randomIntBetween(1, 5)).numberOfReplicas(randomIntBetween(0, 5));
+        IndexMetaData indexMetaData = indexMetadataBuilder.build();
+
+        ClusterState clusterState =
+            ClusterState.builder(emptyClusterState()).metaData(MetaData.builder().put(indexMetaData, true).build()).build();
+
+        try (NoOpClient client = getCreateSnapshotRequestAssertingClient(repository, snapshotName)) {
+            CreateSnapshotStep step = new CreateSnapshotStep(randomStepKey(), randomStepKey(), client);
+            step.performAction(indexMetaData, clusterState, null, new AsyncActionStep.Listener() {
+                @Override
+                public void onResponse(boolean complete) {
```

Comment:
Can add `assertTrue(complete)` here as well (and a `fail` clause in the `onFailure` below)